app/src/components/data-grid/autocomplete/AsyncAutocompleteDataGridEditCell.tsx

```diff
@@ -0,0 +1,204 @@
+import { mdiMagnify } from '@mdi/js';
+import Icon from '@mdi/react';
+import Autocomplete, { createFilterOptions } from '@mui/material/Autocomplete';
+import Box from '@mui/material/Box';
+import CircularProgress from '@mui/material/CircularProgress';
+import TextField from '@mui/material/TextField';
+import { GridRenderCellParams, GridValidRowModel, useGridApiContext } from '@mui/x-data-grid';
+import { IAutocompleteDataGridOption } from 'components/data-grid/autocomplete/AutocompleteDataGrid.interface';
+import { DebouncedFunc } from 'lodash-es';
+import { useEffect, useState } from 'react';
+
+export interface IAsyncAutocompleteDataGridEditCell<
+  DataGridType extends GridValidRowModel,
+  ValueType extends string | number
+> {
+  /**
+   * Data grid props for the cell.
+   *
+   * @type {GridRenderCellParams<DataGridType>}
+   * @memberof IAsyncAutocompleteDataGridEditCell
+   */
+  dataGridProps: GridRenderCellParams<DataGridType>;
+  /**
+   * Function that returns a single option. Used to translate an existing value to its matching option.
+   *
+   * @memberof IAsyncAutocompleteDataGridEditCell
+   */
+  getCurrentOption: (value: ValueType) => Promise<IAutocompleteDataGridOption<ValueType> | null>;
+  /**
+   * Search function that returns an array of options to choose from.
+   *
+   * @memberof IAsyncAutocompleteDataGridEditCell
+   */
+  getOptions: DebouncedFunc<
+    (
+      searchTerm: string,
+      onSearchResults: (searchResults: IAutocompleteDataGridOption<ValueType>[]) => void
+    ) => Promise<void>
+  >;
+}
+
+/**
+ * Data grid single value asynchronous autocomplete component for edit.
+ *
+ * @template DataGridType
+ * @template ValueType
+ * @param {IAsyncAutocompleteDataGridEditCell<DataGridType, ValueType>} props
+ * @return {*}
+ */
+const AsyncAutocompleteDataGridEditCell = <DataGridType extends GridValidRowModel, ValueType extends string | number>(
+  props: IAsyncAutocompleteDataGridEditCell<DataGridType, ValueType>
+) => {
+  const { dataGridProps, getCurrentOption, getOptions } = props;
+
+  const apiRef = useGridApiContext();
+
+  // The current data grid value
+  const dataGridValue = dataGridProps.value;
+  // The input field value
+  const [inputValue, setInputValue] = useState('');
```

Comment:
Should this use `useState<ValueType>('')`?

Reply:
ValueType isn't a string so when placed there it complains that `string` is not assignable to `ValueType | (() => ValueType)`

Reply:
Input value is just the user search string. Technically it should match whatever type the option `label` is, which in all of our autocomplete components is hardcoded to be a string.

Reply:
could change this to `useState<IMultiAutocompleteFieldOption['label']>('');`

Reply:
updated